### ISSUE TITLE
Extreme mutation and lambdas (anonymous functions)

### ISSUE THREAD

**Issue by arnobl** [STAMP-project/pitest-descartes#123]
#### Characteristics
- **Issue Type**: feature

#### Description

Java lambdas are anonymous functions.
As functions, we may consider them as candidates for extreme mutation.
For example

```java
Platform.runLater(() -> {
 ...
});
```
may be replaced by 
```java
Platform.runLater(() -> {});
```

Same things for consumers, suppliers, functions

**Comment by arnobl** [comment]
Difficult points compared to standard extreme mutation:
- since functions are anonymous, we can refer to their name to ignore or not (eg log method)

**Comment by oscarlvp** [comment]
Do you know a way to tell them apart from other compiler-generated methods?
We filter all methods with special names, so lambdas they are given a similar name in the compiled code and therefore they also get filtered.

**Comment by arnobl** [comment]
BTW, an example that illustrates this feature request.

If I write an anonymous class like this one:
```java
disposable = fsm.currentState().subscribe(new Consumer<Map.Entry<OutputState<E>, OutputState<E>>>() {
  @Override
  public void accept(final Map.Entry<OutputState<E>, OutputState<E>> current) throws Exception {
    InteractionImpl.this.updateEventsRegistered(current.getValue(), current.getKey());
   }
 });
```
Descartes creates a mutant for `accept`.

If I turn this anonymous class into an anonymous function, like this one:
```java
disposable = fsm.currentState().subscribe(
    current -> updateEventsRegistered(current.getValue(), current.getKey())
);
```
Descartes does not create a mutant for this anonymous method.

**Comment by arnobl** [comment]
> I must dig into that but I think lambdas have also special names similar to those generated by the compiler.

I do not know. I think that
- lambdas have specific byte-code and may require a process different than method to be processed
- we can start by considering all the lambdas to then look at the name of the function in their interface. However, such names are usually useless as they are generic (`consume`, `apply`).

**Comment by oscarlvp** [comment]
There is a difference in the generated code, anonymous classes are translated into classes while lambdas are translated into methods. For example:

```Java
Supplier<Integer> fromLambda = () -> 4;
```
gets translated to
```
private static java.lang.Integer lambda$main$0();
    Code:
       iconst_4
       invokestatic  #28 
       areturn
```
while 

```Java
Supplier<Integer> fromAClass = new Supplier<Integer>(){
            public Integer get() {
                return 1;
            }
        };
```
gets translated to
```
class Lambdas$1 implements java.util.function.Supplier<java.lang.Integer> {
  Lambdas$1();
    Code:
       0: aload_0
       1: invokespecial #1                  // Method java/lang/Object."<init>":()V
       4: return

  public java.lang.Integer get();
    Code:
       0: iconst_1
       1: invokestatic  #7                  // Method java/lang/Integer.valueOf:(I)Ljava/lang/Integer;
       4: areturn

  public java.lang.Object get();
    Code:
       0: aload_0
       1: invokevirtual #13                 // Method get:()Ljava/lang/Integer;
       4: areturn
}
```

Descartes, discards all methods with `$` on the name, as they are compiler-generated. Compiler generated methods are also marked as `synthetic` by the compiler, but not all compiled generated methods are marked as such.

So, a quick workaround is to allow the mutation in names starting with `lambda$`, if the notation is consistent.

**Comment by arnobl** [comment]
With `+STOP_METHODS(except[synthetic])` Descartes mutates anonymous methods indeed.
But it also creates false positive mutants (strange cases related to inheritance).

> Descartes, discards all methods with $ on the name, as they are compiler-generated.

Where? I can't find that in the code.

**Comment by arnobl** [comment]
anyway. I applied changes in Descartes to spot lambdas.
I will open a PR for discussing code.

**Comment by oscarlvp** [comment]
Solved by #124
